invenio_rdm_records/tokens/resource_access.py

```diff
@@ -0,0 +1,92 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2023 CERN.
+#
+# Invenio-RDM-Records is free software; you can redistribute it and/or modify
+# it under the terms of the MIT License; see LICENSE file for more details.
+
+"""Resource access tokens API."""
+
+from datetime import datetime, timedelta
+
+import jwt
+from flask import current_app
+from invenio_oauth2server.models import Token
+
+from .errors import (
+    ExpiredTokenError,
+    InvalidTokenError,
+    InvalidTokenIDError,
+    MissingTokenIDError,
+)
+from .scopes import tokens_generate_scope
+
+
+class ResourceAccessToken:
+    """Resource access tokens."""
+
+    def __init__(self, token):
+        """Initialize RAT token."""
+        self.token = token
+
+    def validate_rat(self, record, filename):
+        """Decodes a JWT token's payload and signer and performs validation."""
+        # Retrieve token ID from "kid"
+        try:
+            headers = jwt.get_unverified_header(self.token)
+            access_token_id = headers.get("kid")
+        except jwt.InvalidTokenError:
+            raise InvalidTokenError()
+
+        if not access_token_id:
+            raise MissingTokenIDError()
+        if not access_token_id.isdigit():
+            raise InvalidTokenIDError()
+
+        access_token = Token.query.get(int(access_token_id))
+        if (
+            not access_token
+            or tokens_generate_scope.id not in access_token.scopes
+            or (
+                access_token.expires is not None  # expired PAT check
```

Comment:
Added PAT expiration check here as it seems valid to me. Not present in Zenodo implementation though, so feel free to comment on removal

Reply:
If PATs never expire, this check indeed doesn't make sense, thanks for noticing!